dizoo/league_demo/league_scheduler.py

```diff
@@ -0,0 +1,172 @@
+        self.change_range = change_range
+
+        assert isinstance(threshold, float) or isinstance(threshold, int), 'The threshold should be a float/int number'
+        self.threshold = threshold
+
+        assert optimize_mode in ['min', 'max'], 'The optimize_mode should be one of [\'min\', \'max\']'
+        self.optimize_mode = optimize_mode
+
+        assert isinstance(patience, int), 'The patience should be a integer greater than or equal to 0'
+        assert patience >= 0, 'The patience should be a integer greater than or equal to 0'
+        self.patience = patience
+
+        assert isinstance(cooldown, int), 'The cooldown_counter should be a integer greater than or equal to 0'
+        assert cooldown >= 0, 'The cooldown_counter should be a integer greater than or equal to 0'
+        self.cooldown = cooldown
+        self.cooldown_counter = cooldown
+
+        self.last_metrics = None
+        self.bad_epochs_num = 0
+        
+
+    def step(self, metrics, cfg):
+        assert isinstance(metrics, float), 'The metrics should be a float number'
+        cur_metrics = metrics
+        
+        if self.is_better(cur_metrics):
+            self.bad_epochs_num = 0
+        else:
+            self.bad_epochs_num += 1
+        self.last_metrics = cur_metrics
+        
+        if self.in_cooldown:
+            self.cooldown_counter -= 1
+            self.bad_epochs_num = 0  # ignore any bad epochs in cooldown
+        
+        if self.bad_epochs_num > self.patience:
+            self.update_para(cfg)
+            self.cooldown_counter = self.cooldown
+            self.bad_epochs_num = 0
+            
+    
+    def update_para(self, cfg):
+        if self.schedule_parameter == 'entropy_weight':
+            entropy_weight = cfg.policy.learn.entropy_weight
+            if self.schedule_mode == 'reduce':
+                cfg.policy.learn.entropy_weight = max(entropy_weight - self.factor, self.change_range[0])
+            
+            if self.schedule_mode == 'add':
+                cfg.policy.learn.entropy_weight = min(entropy_weight + self.factor, self.change_range[1])
+            
+            if self.schedule_mode == 'multi':
+                if self.factor >= 1:
+                    cfg.policy.learn.entropy_weight = min(self.factor*entropy_weight, self.change_range[1])
+                else:
+                    cfg.policy.learn.entropy_weight = max(self.factor*entropy_weight, self.change_range[0])
+            
+            if self.schedule_mode == 'div':
+                if self.factor >= 1:
+                    cfg.policy.learn.entropy_weight = max(float(entropy_weight) / float(self.factor), self.change_range[0])
+                else:
+                    cfg.policy.learn.entropy_weight = min(float(entropy_weight) / float(self.factor), self.change_range[1])
+
+        if self.schedule_parameter == 'learning_rate':
```

Comment:
You can simplify these `if-else` codes with table-driven method and function abstraction:
```python
schedule_fn = {
    'reduce': lambda x, y, z: max(x - y, z[0]),
    'add': lambda x, y, z: min(x + y, z[1]),
    'multi': lambda x, y, z: min(x * y, z[1]) if y >= 1 else max(x * y, z[0]),
    'div': lambda x, y, z: max(x / y, z[0]) if y >= 1 else min(x / y, z[1]),
}
schedule_mode_list = list(schedule_fn.keys())

if self.schedule_mode in schedule_mode_list:
    cfg.policy.learn.learning_rate = schedule_fn(cfg.policy.learn.learning_rate, self.factor, self.change_range)
else:
    raise KeyError("invalid schedule_mode({}) in {}".format(self.schedule_mode, schedule_mode_list))
```

Reply:
fixed.